R/test-that.R

```diff
@@ -13,6 +13,46 @@
 #' changes. Avoid changing global state, when possible, and reverse any changes
 #' that you do make.
 #'
+#' # Detecting changes to global state
+#'
+#' One of the most pernicious challenges to debug is when a test runs fine
+#' in your test suite, but fails when you run it interactively (or similarly,
+#' it fails randomly when running your tests in parallel). One of the most
+#' common causes of this problem is accidentally changing global state in a
+#' previous test (e.g. changing an option, an environment variable, or the
+#' working directory). This is hard to debug, because it's very hard to figure
+#' out which test made the change.
+#'
+#' Luckily testthat provides a tool to figure out if tests are changing global
```

Comment:
It's not clear from this documentation where you should define this function. Within a test, outside a test, within a helper file, in the `testthat.R` file?

I see in the implementation that you're looking in `testthat_env` so I guess the best place would be a helper file for this to take global effect?

Is this a common pattern that we use to implement hooks? I think I would have expected the UI to be something like:

```r
# Enable
testthat::register_state_watch(function() list())

# Disable? Then need to check for `NULL` in `after`
testthat::register_state_watch(NULL)
```

Then as a user I'd feel more comfortable registering a watcher because I know it will take effect after the call. This makes it easier to wrap in another function as well, perhaps exported by a different package.

Reply:
Oh I like that API much better.

Reply:
Maybe `set_state_watcher()` since that at least somewhat parallels `set_max_fails()`.